src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -1064,51 +1067,106 @@ private static ITypeReference GetArrayItemType(ArrayAccessSyntax syntax, IDiagno
+        {
+            baseType = UnwrapType(baseType);
+
+            switch (baseType)
             {
-                var errors = new List<ErrorDiagnostic>();
+                case ErrorType:
+                    return baseType;
+                case TypeSymbol _ when baseType.GetDiagnostics().Any():
+                    return ErrorType.Create(baseType.GetDiagnostics());
+                case ObjectType objectType:
+                    if (!syntax.PropertyName.IsValid)
+                    {
+                        // the property is not valid
+                        // there's already a parse error for it, so we don't need to add a type error as well
+                        return ErrorType.Empty();
+                    }
 
-                var baseType = typeManager.GetTypeInfo(syntax.BaseExpression);
-                CollectErrors(errors, baseType);
+                    return TypeHelper.GetNamedPropertyType(objectType, syntax.PropertyName, syntax.PropertyName.IdentifierName, syntax.SafeAccessMarker is not null || TypeValidator.ShouldWarn(objectType), diagnostics);
 
-                if (PropagateErrorType(errors, baseType))
-                {
-                    return ErrorType.Create(errors);
-                }
+                case DiscriminatedObjectType _:
+                    // TODO: We might be able use the declared type here to resolve discriminator to improve the assigned type
+                    return LanguageConstants.Any;
 
-                baseType = UnwrapType(baseType);
+                case TypeSymbol _ when TypeValidator.AreTypesAssignable(baseType, LanguageConstants.Object):
+                    // We can assign to an object, but we don't have a type for that object.
+                    // The best we can do is allow it and return the 'any' type.
+                    return LanguageConstants.Any;
 
-                switch (baseType)
-                {
-                    case ObjectType objectType:
-                        if (!syntax.PropertyName.IsValid)
-                        {
-                            // the property is not valid
-                            // there's already a parse error for it, so we don't need to add a type error as well
-                            return ErrorType.Empty();
-                        }
+                default:
+                    // can only access properties of objects
+                    return InvalidAccessExpression(DiagnosticBuilder.ForPosition(syntax.PropertyName).ObjectRequiredForPropertyAccess(baseType), diagnostics, syntax.SafeAccessMarker is not null);
+            }
+        }
 
-                        return TypeHelper.GetNamedPropertyType(objectType, syntax.PropertyName, syntax.PropertyName.IdentifierName, TypeValidator.ShouldWarn(objectType), diagnostics);
+        private TypeSymbol GetAccessedType(AccessExpressionSyntax syntax, IDiagnosticWriter diagnostics)
+        {
+            Stack<AccessExpressionSyntax> chainedAccesses = new();
```

Comment:
Could you add an explanation here why this algorithm is necessary (vs the regular recursion we're using elsewhere in this visitor)?